This is my implementation of a JSON cache

live627 opened the pull request:
Fixes #5826

- encodes the data with the JSON lib 
- removes the risky code evaluation that sometimes threw fatal errors
- use a reader lock to avoid reading exclusively locked files
- don't block the script if any lock could not be obtained

[A user comment on php.net about how file locks work](https://www.php.net/manual/en/function.flock.php#78318)

jdarwood007 commented on Sources/CacheAPI-smf.php:
```diff
@@ -49,67 +49,107 @@ public function isSupported($test = false)
-
-			if (function_exists('apc_delete_file'))
-				@apc_delete_file($cachedir . '/data_' . $key . '.php');
-
```
I think we still need these.  On systems where this would take effect, you have made it so it would keep rebuilding the cache file until it catches.  So in effect it could run the code it should be skipping dozens of times before the opcache/apc clears out.  To help simplify this, we could drop it into a sub method.

I understand we don't do it when clearing the entire cache folder.  Something should be done for that as well.

live627 replied:
These calls are no longer needed because these files are no longer compiled and cached as bytecode

jdarwood007 replied:
Your right.  I was thinking it still gets cached, but the only one would be the file system caching it.  We still may want to call clearstatcache() then in order to inform php to no longer cache it on that side.

sbulen commented:
This tests OK in my tests - it definitely solves the original problem.  Two notes:
 - The reads return NULL sometimes, for both wamp & linux (due to...)
 - I don't think you want the LOCK_NBs

LOCK_NB causes execution to continue even if the file is unavailable.  For reads, since the return value is checked, what ends up happening is it is treated as a failed read.  This is why there are occasional NULLs returned.  Note the calling function will detect the lack of a cache & rebuild the cache unnecessarily.  I think the wait is preferred, i.e., no LOCK_NB, so the read is solid & the cache is used.

On the writes, with the LOCK_NB, you are going to return false & attempt to delete whatever was out there.  It will require a subsequent call to update the cache.  I think the wait is preferred, i.e., no LOCK_NB, because it will only take one update.

If you remove the LOCK_NBs, it will wait as normal & complete the read or write.  I have confirmed you get no NULLs if you remove them.  At that point, however, you have basically rewritten file_put_contents & file_get_contents...  

I think it would be better to remove the LOCK_NBs ~~& just use file_put_contents & file_get_contents~~ (edit: see below).  It's a bit more solid & simple.

sbulen commented:
I've been doing more testing...

I had only tested the _modified_ PR on linux earlier...  I just tested it on wamp now and it is perfect on wamp as well as linux - zero NULLs.  _But ONLY if you get rid of the LOCK_NBs..._  So, if modified, it may in fact have an advantage over file_put_contents & file_get_contents (which are perfect on linux but occasionally return NULLs on wamp).  I don't know why, but it does.

I'm still testing, but so far, it looks good!

live627 committed:
Revert "look ma, no locks! "

This reverts commit 3d4b41bd3eb4a8ff71fbe790db3cd598e56a6988.

live627 commented:
I think I may have another approach which is free of locks.

albertlast commented:
Until php 8.0 json function can be disabled right?

live627 commented:
SMF 2.1 does JSON everywhere

On Fri, Aug 7, 2020 at 1:16 PM albertlast <notifications@github.com> wrote:

> Until php 8.0 json function can be disabled right?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/SimpleMachines/SMF2.1/pull/6211#issuecomment-670695761>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADJNN3KFXTAAXLOLPDVKTLR7ROKNANCNFSM4PSQZDWA>
> .
>

albertlast commented:
i forgot, right

sbulen commented:
I tested the latest, and it has the same issues as Suki's version:  The cache PUT collisions take place on the rename:
![image](https://user-images.githubusercontent.com/23568484/89740739-4a727a80-da40-11ea-9c83-8969b7a1b056.png)

This happens on lamp, linux seems to behave better.

As noted earlier, what worked PERFECT was the prior version of this PR, but without the LOCK_NBs (but keeping the other locks).

live627 committed:
block while acquiring lock

sbulen commented:
Just retested, using the latest & greatest, & this PR now has 100% success on PUTs & GETs, across wamp & linux as well!  

This has actually worked out well, learning what worked & what didn't across multiple PRs, trying different approaches to the same problem.

live627 commented:
ok then I'll now leave it as is since it works.

Just some documentation left....

On Sun, Aug 9, 2020 at 7:09 PM sbulen <notifications@github.com> wrote:

> Just retested, using the latest & greatest, & this PR now has 100% success
> on PUTs & GETs, across wamp & linux as well!
>
> This has actually worked out well, learning what worked & what didn't
> across multiple PRs, trying different approaches to the same problem.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/SimpleMachines/SMF2.1/pull/6211#issuecomment-671134908>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADJNN7OQWWGSLIPJFFBX5TR75JD5ANCNFSM4PSQZDWA>
> .
>

live627 committed:
This is my implementation of a JSON cache

Fixes #5847

- encodes the data with the JSON lib
- removes the risky code evaluation that sometimes threw fatal errors
- use a reader lock to avoid reading exclusively locked files
- don't block the script if any lock could not be obtained

live627 committed:
look ma, no locks!

live627 commented:
The rebase was worth it!

On Fri, Nov 6, 2020 at 3:52 PM Michel Mendiola <notifications@github.com>
wrote:

> *@MissAllSunday* approved this pull request.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/SimpleMachines/SMF2.1/pull/6211#pullrequestreview-525543065>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADJNN47C3Z25DERDADTPE3SOR44XANCNFSM4PSQZDWA>
> .
>
